### PR TITLE
Add mulitpath support to iscsi plugin

### PR DESCRIPTION
**Pull request by humblec**
#issue https://github.com/kubernetes/kubernetes/issues/39345
Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

### CODE REVIEW

**Comment by humblec** [comment]
@rootfs PTAL.  I havent updated the references ( `update-all.sh`)  in apis. Once this looks good, I will add it as a seperate commit.

**Comment by rootfs** [comment]
@k8s-bot ok to test

File: pkg/api/types.go [hunk 1]
```diff
@@ -603,6 +603,10 @@ type ISCSIVolumeSource struct {
 	// the portal is either an IP or ip_addr:port if port is other than default (typically TCP ports 860 and 3260)
 	// +optional
 	TargetPortal string
+	// Required: list of iSCSI target portal ips
+	// the portal is either an IP or ip_addr:port if port is other than default (typically TCP ports 860 and 3260)
+	// +optional
+	BackupPortal []string
```
Comment (rootfs) [inline pkg/api/types.go]: inconsistent with those defined in `pkg/api/v1/types.go`

Comment (humblec) [inline pkg/api/types.go]: done.

**Comment by rootfs** [comment]
```
W0116 21:02:34.943] errors in package "k8s.io/kubernetes/pkg/generated/openapi":
W0116 21:02:34.944] output for "openapi/zz_generated.openapi.go" differs; first existing/expected diff: 
W0116 21:02:34.944]   "iqn\": {\n\t\t\t\t\t\tSchemaProps: spec.SchemaProps{\n\t\t\t\t\t\t\tDescription: \"Target iSCSI Qualified Name.\",\n\t\t\t"
W0116 21:02:34.944]   "backupPortal\": {\n\t\t\t\t\t\tSchemaProps: spec.SchemaProps{\n\t\t\t\t\t\t\tDescription: \"iSCSI target portal List."
W0116 21:02:34.944] 
W0116 21:02:34.975] make[1]: *** [verify_gen_openapi] Error 1
W0116 21:02:34.976] make: *** [verify_generated_files] Error 2
I0116 21:02:35.076] Makefile.generated_files:502: recipe for target 'verify_gen_openapi' failed
```

File: pkg/api/v1/types.go [hunk 2]
```diff
@@ -965,6 +965,9 @@ type ISCSIVolumeSource struct {
 	// iSCSI target portal. The portal is either an IP or ip_addr:port if the port
 	// is other than default (typically TCP ports 860 and 3260).
 	TargetPortal string `json:"targetPortal" protobuf:"bytes,1,opt,name=targetPortal"`
+	// iSCSI target portal List. The portal is either an IP or ip_addr:port if the port
+	// is other than default (typically TCP ports 860 and 3260).
+	BackupPortal []string `json:"backupPortal" protobuf:"bytes,7,opt,name=backupPortal"`
```
Comment (rootfs) [inline pkg/api/v1/types.go]: reorder the new param

File: pkg/volume/iscsi/iscsi_util.go [hunk 3]
```diff
@@ -103,8 +103,9 @@ func (util *ISCSIUtil) MakeGlobalPDName(iscsi iscsiDisk) string {
 
 func (util *ISCSIUtil) AttachDisk(b iscsiDiskMounter) error {
 	var devicePath string
+	var devicePaths []string
 	var iscsiTransport string
-
+	reqExist := false
```
Comment (rootfs) [inline pkg/volume/iscsi/iscsi_util.go]: you can skip this boolean by check if `len(devicePaths) > 0`

File: pkg/volume/iscsi/iscsi_util.go [hunk 4]
```diff
@@ -113,33 +114,52 @@ func (util *ISCSIUtil) AttachDisk(b iscsiDiskMounter) error {
...
+				continue
+			}
+			exist = waitForPathToExist(devicePath, 10, iscsiTransport)
+			if !exist {
+				glog.Errorf("Could not attach disk: Timeout after 10s")
+			} else {
+				devicePaths = append(devicePaths, devicePath)
+			}
+		} else {
+			glog.Infof("iscsi: devicepath (%s) exist", devicePath)
```
Comment (rootfs) [inline pkg/volume/iscsi/iscsi_util.go]: glog.V(4)
exist -> exists

File: pkg/volume/iscsi/iscsi_util.go [hunk 5]
```diff
@@ -113,33 +114,52 @@ func (util *ISCSIUtil) AttachDisk(b iscsiDiskMounter) error {
...
+			devicePaths = append(devicePaths, devicePath)
+			reqExist = exist
 		}
+
+	}
+
+	if !reqExist {
```
Comment (rootfs) [inline pkg/volume/iscsi/iscsi_util.go]: len(devicePaths) == 0

File: pkg/volume/iscsi/iscsi_util.go [hunk 6]
```diff
@@ -153,9 +173,15 @@ func (util *ISCSIUtil) AttachDisk(b iscsiDiskMounter) error {
 		return err
 	}
 
-	// check if the dev is using mpio and if so mount it via the dm-XX device
-	if mappedDevicePath := b.deviceUtil.FindMultipathDeviceForDevice(devicePath); mappedDevicePath != "" {
-		devicePath = mappedDevicePath
+	for _, path := range devicePaths {
+		if path == "" {
```
Comment (rootfs) [inline pkg/volume/iscsi/iscsi_util.go]: why empty path exists?

File: pkg/volume/iscsi/iscsi_util.go [hunk 7]
```diff
@@ -113,33 +114,52 @@ func (util *ISCSIUtil) AttachDisk(b iscsiDiskMounter) error {
...
+
+	}
+
+	if !reqExist {
+		glog.Errorf("iscsi: failed to get any path for iscsi disk")
+		return errors.New("failed to get any path for iscsi disk")
 	}
+
+	//Make sure we use a valid devicepath to find mpio device.
+	devicePath = devicePaths[0]
```
Comment (rootfs) [inline pkg/volume/iscsi/iscsi_util.go]: if empty path exists, this is not safe!

**Review by rootfs — CHANGES_REQUESTED** [review]

**Comment by rootfs** [comment]
this needs release note update too,

File: pkg/volume/iscsi/iscsi_util.go (hunk 6 above)
Comment (humblec) [inline pkg/volume/iscsi/iscsi_util.go]: It wont or very very rare chance, that check is added while considering the scenario where admin logged out from one path in between our fetch and use of that path.

File: pkg/volume/iscsi/iscsi_util.go (hunk 7 above)
Comment (humblec) [inline pkg/volume/iscsi/iscsi_util.go]: devicePath array entry will not be empty. The devicepath would be  filled with a valid path in any case. It shouldnt be empty path as we are populating the array by validating the path.

File: pkg/volume/iscsi/iscsi_util.go (hunk 5 above)
Comment (humblec) [inline pkg/volume/iscsi/iscsi_util.go]: sure. Removed the flag.

File: pkg/volume/iscsi/iscsi_util.go (hunk 4 above)
Comment (humblec) [inline pkg/volume/iscsi/iscsi_util.go]: Done.

File: pkg/volume/iscsi/iscsi_util.go (hunk 3 above)
Comment (humblec) [inline pkg/volume/iscsi/iscsi_util.go]: Done.

**Comment by rootfs** [comment]
have you run `hack/verify-openapi-spec.sh`

**Comment by tobad357** [comment]
Looking at the logic the iscsi implementation will always try the BackupPortal event if the primary Portal succeeds. This is something which I think is the best option, however if it always uses BackupPortal then it isn't really a Backup.

Wouldn't it be better to have a list for configurations which would be called Portals
If the user configures both Portal and Portals then just append Portal to Portals
Validation should be that either Portal or Portals has been configured

**Comment by humblec** [comment]
@tobad357 Thanks for the comment. The implementation has been done by considering the suggestions made in issue # https://github.com/kubernetes/kubernetes/issues/21306 .  If everyone think for a different name than `backupportal`, I can definitely change it.

**Comment by tobad357** [comment]
@humblec When I wrote that issue my thinking was more in line with try the first portal and if it fails then try the backup. As it now is a list of portals and all will be used I vote to call the list "portals" and keep "portal" for backwards compatibility and ease of use

**Comment by humblec** [comment]
@rootfs `hack/verify-openapi-spec.sh` gives error. Looking into it.

**Comment by rootfs** [comment]
@k8s-bot verify test this

**Comment by humblec** [comment]
Looks like now, `codecgen` is going wrong, checking

**Comment by humblec** [comment]
@rootfs the ` ./hack/verify-bazel.sh` works fine at my end, however failing here. Not sure, what is missing.  do you have any pointer ?

**Comment by humblec** [comment]
`upate-all.sh` and each individual verification scripts works perfectly in local copy. However it fails here.  

Also the errors on 'generated go files` looks really awkward to me.

~~~
W0124 16:50:10.657] cmd/kubernetes-discovery/pkg/apis/apiregistration/zz_generated.deepcopy.go:46: undefined: "k8s.io/kubernetes/pkg/conversion".Cloner
W0124 16:50:10.658] cmd/kubernetes-discovery/pkg/apis/apiregistration/zz_generated.deepcopy.go:61: undefined: "k8s.io/kubernetes/pkg/conversion".Cloner
W0124 16:50:10.658] cmd/kubernetes-discovery/pkg/apis/apiregistration/zz_generated.deepcopy.go:79: undefined: "k8s.io/kubernetes/pkg/conversion".Cloner
~~~

**Comment by humblec** [comment]
@rootfs  Finallly all the tests are passed 👍 .  I am squashing the commits.

**Comment by rootfs** [comment]
@k8s-bot test this

File: pkg/api/v1/types.go [hunk 8]
```diff
@@ -988,6 +988,9 @@ type ISCSIVolumeSource struct {
 	// Defaults to false.
 	// +optional
 	ReadOnly bool `json:"readOnly,omitempty" protobuf:"varint,6,opt,name=readOnly"`
+	// iSCSI target portal List. The portal is either an IP or ip_addr:port if the port
+	// is other than default (typically TCP ports 860 and 3260).
```
Comment (rootfs) [inline pkg/api/v1/types.go]: `+optional`

**Comment by rootfs** [comment]
@humblec @tobad357 let use `portals` for the new API object. The logic is to make a string array, starting from `portal` and followed by those in `portals`. 

One thing I haven't figured out is that what if `portal` and new `portals` happen to map to different LUNs (vs. the same mpio LUN)? Is there any way to identify this misconfiguration?

**Review by rootfs — CHANGES_REQUESTED** [review]

**Comment by humblec** [comment]
@rootfs @tobad357 sure. I can `s/backupPortal/portals` at user input side and keep the internal reference in the code as is.  
Regarding  multiportal with different lun numbers: our current volume spec has to be considered/projected  as a spec for single `IQN` with choices of having `>=1` portal and `single lun ` number, which is the most common method afaict.  As we login using the `IQN`, I feel  it should actually avoid  the misconfigurations.  If we want to support more combinations ( multiple portals with different lun numbers..etc) we have to really think about and get more details about the setups then consider a common solution. As per me, the further improvements  can be considered later after analysing different scenarios. Atleast multiple TP support looks to be an immediate ask from many users.

**Comment by rootfs** [comment]
@humblec sounds good!

**Comment by humblec** [comment]
Thanks @rootfs , I am replacing `backupPortal with portals`, will update the PR.

File: pkg/api/v1/types.go (hunk 8 above)
Comment (humblec) [inline pkg/api/v1/types.go]: Done.

**Comment by humblec** [comment]
@rootfs  comments are addressed and  tests are passed. 👍

File: pkg/volume/iscsi/iscsi.go [hunk 9]
```diff
@@ -163,6 +167,7 @@ type iscsiDisk struct {
 	volName string
 	podUID  types.UID
 	portal  string
```
Comment (rootfs) [inline pkg/volume/iscsi/iscsi.go]: Since it is an array now, you can consolidate `portal` and `portals` into `portals`.

**Comment by rootfs** [comment]
@humblec thanks! almost there, one last comment

**Comment by humblec** [comment]
@rootfs rebased and also addressed your  comment. The AWS failure looks to be spurious one. Can you please review/merge this?

File: federation/apis/swagger-spec/extensions_v1beta1.json [hunk 10]
```diff
@@ -7063,6 +7070,13 @@
        "type": "string"
       },
       "description": "verbs is a list of supported kube verbs (this includes get, list, watch, create, update, patch, delete, deletecollection, and proxy)"
+     },
+     "shortNames": {
```
Comment (rootfs) [inline federation/apis/swagger-spec/extensions_v1beta1.json]: where does this change come from?

Comment (humblec) [inline federation/apis/swagger-spec/extensions_v1beta1.json]: @rootfs looks to be from `update-federation-swagger-spec`.

**Comment by humblec** [comment]
@rootfs once the change looks good, I will update the spec.

File: examples/volumes/iscsi/README.md [hunk 11]
```diff
@@ -18,7 +18,7 @@ If there isn't one in place then it is possible to setup a software version on L
 
 ## Creating the pod with iSCSI persistent storage
 
-Once you have configured the iSCSI initiator, you can create a pod based on the example *iscsi.yaml*. In the pod YAML, you need to provide *targetPortal* (the iSCSI target's **IP** address and *port* if not the default port 3260), target's *iqn*, *lun*, and the type of the filesystem that has been created on the lun, and *readOnly* boolean. No initiator information is required.
+Once you have configured the iSCSI initiator, you can create a pod based on the example *iscsi.yaml*. In the pod YAML, you need to provide *targetPortal* (the iSCSI target's **IP** address and *port* if not the default port 3260), target's *iqn*, *lun*, and the type of the filesystem that has been created on the lun, and *readOnly* boolean. No initiator information is required. If you have more than one target portal for a single IQN, you can mention other portal IPs in *portals* field.
```
Comment (rootfs) [inline examples/volumes/iscsi/README.md]: "more than one target portals"

**Comment by humblec** [comment]
@childsb looks like this is not going to submit queue, do we need to trigger `@k8s-bot node e2e test this` ?

**Comment by humblec** [comment]
@childsb looks like my last comment triggered the failed test and it has passed now. Is there any special flag ( `approved` )  required on this PR to get this merged?

**Comment by humblec** [comment]
@k8s-bot gci gce e2e test this

**Comment by humblec** [comment]
@rootfs can u restore LGTM and also do the needful for getting this merged ?

**Comment by rootfs** [comment]
@humblec please rebase and i'll lgtm

**Comment by humblec** [comment]
@rootfs rebased and tests are green. Can you please restore LGTM and provide `approved` flag ?

**Comment by humblec** [comment]
@rootfs thanks! @childsb can you please put `approved` flag on this PR ?

**Pull request merged**